Flaky-test: ElasticSearchClientTests.testBulkRetry

ElasticSearchClientTests is flaky. The testBulkRetry test method fails sporadically.

[example failure](https://github.com/apache/pulsar/pull/11564/checks?check_run_id=3290485512)


<details>
<summary>Full exception stacktrace</summary>
<code><pre>
Error:  Tests run: 8, Failures: 1, Errors: 0, Skipped: 6, Time elapsed: 145.874 s <<< FAILURE! - in org.apache.pulsar.io.elasticsearch.ElasticSearchClientTests
Error:  testBulkRetry(org.apache.pulsar.io.elasticsearch.ElasticSearchClientTests)  Time elapsed: 1.162 s  <<< FAILURE!
java.lang.AssertionError: expected [2] but found [3]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:797)
	at org.testng.Assert.assertEquals(Assert.java:807)
	at org.apache.pulsar.io.elasticsearch.ElasticSearchClientTests.testBulkRetry(ElasticSearchClientTests.java:184)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)

[INFO] 
[INFO] Results:
[INFO] 
Error:  Failures: 
Error:  org.apache.pulsar.io.elasticsearch.ElasticSearchClientTests.testBulkRetry(org.apache.pulsar.io.elasticsearch.ElasticSearchClientTests)
[INFO]   Run 1: PASS
Error:    Run 2: ElasticSearchClientTests.testBulkRetry:184 expected [2] but found [3]
[INFO] 
[INFO] 
Error:  Tests run: 49, Failures: 1, Errors: 0, Skipped: 5
</pre></code>
</details>

another failure

```
Error:  Tests run: 8, Failures: 1, Errors: 0, Skipped: 6, Time elapsed: 134.905 s <<< FAILURE! - in org.apache.pulsar.io.elasticsearch.ElasticSearchClientTests
Error:  testBulkRetry(org.apache.pulsar.io.elasticsearch.ElasticSearchClientTests)  Time elapsed: 1.156 s  <<< FAILURE!
java.lang.AssertionError: expected [2] but found [3]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:797)
	at org.testng.Assert.assertEquals(Assert.java:807)
	at org.apache.pulsar.io.elasticsearch.ElasticSearchClientTests.testBulkRetry(ElasticSearchClientTests.java:184)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

at https://github.com/apache/pulsar/pull/11497/checks?check_run_id=3288037630#step:9:27002

testBulkBlocking is also flaky

```
Error:  Tests run: 8, Failures: 1, Errors: 0, Skipped: 7, Time elapsed: 93.14 s <<< FAILURE! - in org.apache.pulsar.io.elasticsearch.ElasticSearchClientTests
Error:  testBulkBlocking(org.apache.pulsar.io.elasticsearch.ElasticSearchClientTests)  Time elapsed: 10.239 s  <<< FAILURE!
org.awaitility.core.ConditionTimeoutException: Assertion condition defined as a lambda expression in org.apache.pulsar.io.elasticsearch.ElasticSearchClientTests that uses org.apache.pulsar.io.elasticsearch.ElasticSearchClientTests$MockRecord, org.apache.pulsar.io.elasticsearch.ElasticSearchClientTests$MockRecordorg.apache.pulsar.io.elasticsearch.ElasticSearchClient expected [5] but found [4] within 10 seconds.
	at org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:165)
	at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:119)
	at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:31)
	at org.awaitility.core.ConditionFactory.until(ConditionFactory.java:895)
	at org.awaitility.core.ConditionFactory.untilAsserted(ConditionFactory.java:679)
	at org.apache.pulsar.io.elasticsearch.ElasticSearchClientTests.testBulkBlocking(ElasticSearchClientTests.java:222)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.AssertionError: expected [5] but found [4]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.io.elasticsearch.ElasticSearchClientTests.lambda$testBulkBlocking$5(ElasticSearchClientTests.java:223)
	at org.awaitility.core.AssertionCondition.lambda$new$0(AssertionCondition.java:53)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:222)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:209)
	... 4 more
```
in https://github.com/apache/pulsar/pull/11497/checks?check_run_id=3288037630#step:9:38776

PTAL @eolivelli @vroyer

--- a/pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/ElasticSearchClient.java
+++ b/pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/ElasticSearchClient.java
@@ -48,6 +48,7 @@ import org.apache.pulsar.client.api.schema.GenericObject;
 import org.apache.pulsar.functions.api.Record;
 import org.elasticsearch.action.DocWriteRequest;
 import org.elasticsearch.action.DocWriteResponse;
+import org.elasticsearch.action.admin.indices.delete.DeleteIndexRequest;
 import org.elasticsearch.action.admin.indices.refresh.RefreshRequest;
 import org.elasticsearch.action.bulk.BulkItemResponse;
 import org.elasticsearch.action.bulk.BulkProcessor;
@@ -124,7 +125,7 @@ public class ElasticSearchClient {
     final AtomicReference<Exception> irrecoverableError = new AtomicReference<>();
     final ScheduledExecutorService executorService;
 
-    ElasticSearchClient(ElasticSearchConfig elasticSearchConfig) throws MalformedURLException {
+    ElasticSearchClient(ElasticSearchConfig elasticSearchConfig) {
         this.config = elasticSearchConfig;
         this.configCallback = new ConfigCallback();
         this.backoffRetry = new RandomExponentialRetry(elasticSearchConfig.getMaxRetryTimeInSec());
@@ -470,6 +471,16 @@ public class ElasticSearchClient {
                         .indices(indexName)
                         .source(new SearchSourceBuilder().query(QueryBuilders.matchAllQuery())),
                 RequestOptions.DEFAULT);
+    }
+
+    @VisibleForTesting
+    protected org.elasticsearch.action.admin.indices.refresh.RefreshResponse refresh(String indexName) throws IOException {
+        return client.indices().refresh(new RefreshRequest(indexName), RequestOptions.DEFAULT);
+    }
+
+    @VisibleForTesting
+    protected org.elasticsearch.action.support.master.AcknowledgedResponse delete(String indexName) throws IOException {
+        return client.indices().delete(new DeleteIndexRequest(indexName), RequestOptions.DEFAULT);
     }
 
     private <T> T retry(Callable<T> callable, String source) {

--- a/pulsar-io/elastic-search/src/test/java/org/apache/pulsar/io/elasticsearch/ElasticSearchClientTests.java
+++ b/pulsar-io/elastic-search/src/test/java/org/apache/pulsar/io/elasticsearch/ElasticSearchClientTests.java
@@ -31,32 +31,26 @@ import org.testng.annotations.Test;
 
 import java.io.IOException;
 import java.util.Optional;
+import java.util.UUID;
 
 import static org.junit.Assert.*;
 import static org.testng.Assert.assertEquals;
 import static org.testng.Assert.assertNotNull;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.*;
 
 @Slf4j
 public class ElasticSearchClientTests {
 
     public static final String ELASTICSEARCH_IMAGE = Optional.ofNullable(System.getenv("ELASTICSEARCH_IMAGE"))
             .orElse("docker.elastic.co/elasticsearch/elasticsearch-oss:7.10.2-amd64");
-    public final static String INDEX = "myindex";
 
     static ElasticsearchContainer container;
-    static ElasticSearchConfig config;
-    static ElasticSearchClient client;
 
     @BeforeClass
     public static final void initBeforeClass() throws IOException {
         container = new ElasticsearchContainer(ELASTICSEARCH_IMAGE);
         container.start();
-
-        config = new ElasticSearchConfig();
-        config.setElasticSearchUrl("http://" + container.getHttpHostAddress());
-        config.setIndexName(INDEX);
-
-        client = new ElasticSearchClient(config);
     }
 
     @AfterClass
@@ -86,29 +80,50 @@ public class ElasticSearchClientTests {
 
     @Test
     public void testIndexDelete() throws Exception {
-        client.createIndexIfNeeded(INDEX);
-        MockRecord<GenericObject> mockRecord = new MockRecord<>();
-        client.indexDocument(mockRecord, Pair.of("1","{ \"a\":1}"));
-        assertEquals(mockRecord.acked, 1);
-        assertEquals(mockRecord.failed, 0);
-        assertEquals(client.totalHits(INDEX), 1);
+        String index = "myindex-" + UUID.randomUUID();
+        ElasticSearchClient client = new ElasticSearchClient(new ElasticSearchConfig()
+                .setElasticSearchUrl("http://" + container.getHttpHostAddress())
+                .setIndexName(index)
+        );
+        assertTrue(client.createIndexIfNeeded(index));
+        try {
+            MockRecord<GenericObject> mockRecord = new MockRecord<>();
+            client.indexDocument(mockRecord, Pair.of("1", "{ \"a\":1}"));
+            assertEquals(mockRecord.acked, 1);
+            assertEquals(mockRecord.failed, 0);
+            assertEquals(client.totalHits(index), 1);
 
-        client.deleteDocument(mockRecord, "1");
-        assertEquals(mockRecord.acked, 2);
-        assertEquals(mockRecord.failed, 0);
-        assertEquals(client.totalHits(INDEX), 0);
+            client.deleteDocument(mockRecord, "1");
+            assertEquals(mockRecord.acked, 2);
+            assertEquals(mockRecord.failed, 0);
+            assertEquals(client.totalHits(index), 0);
+        } finally {
+            client.delete(index);
+        }
     }
 
     @Test
     public void testIndexExists() throws IOException {
-        assertFalse(client.indexExists("mynewindex"));
-        assertTrue(client.createIndexIfNeeded("mynewindex"));
-        assertTrue(client.indexExists("mynewindex"));
-        assertFalse(client.createIndexIfNeeded("mynewindex"));
+        String index = "mynewindex-" + UUID.randomUUID();
+        ElasticSearchClient client = new ElasticSearchClient(new ElasticSearchConfig()
+                .setElasticSearchUrl("http://" + container.getHttpHostAddress())
+                .setIndexName(index)
+        );
+        assertFalse(client.indexExists(index));
+        assertTrue(client.createIndexIfNeeded(index));
+        try {
+            assertTrue(client.indexExists(index));
+            assertFalse(client.createIndexIfNeeded(index));
+        } finally {
+            client.delete(index);
+        }
     }
 
     @Test
-    public void testTopicToIndexName() {
+    public void testTopicToIndexName() throws IOException {
+        ElasticSearchClient client = new ElasticSearchClient(new ElasticSearchConfig()
+                .setElasticSearchUrl("http://" + container.getHttpHostAddress())
+        );
         assertEquals(client.topicToIndexName("data-ks1.table1"),"data-ks1.table1");
         assertEquals(client.topicToIndexName("persistent://public/default/testesjson"), "testesjson");
         assertEquals(client.topicToIndexName("default/testesjson"), "testesjson");
@@ -123,7 +138,7 @@ public class ElasticSearchClientTests {
 
     @Test
     public void testMalformedDocFails() throws Exception {
-        String index = "indexmalformed";
+        String index = "indexmalformed-" + UUID.randomUUID();
         ElasticSearchConfig config = new ElasticSearchConfig()
                 .setElasticSearchUrl("http://"+container.getHttpHostAddress())
                 .setIndexName(index)
@@ -145,7 +160,7 @@ public class ElasticSearchClientTests {
 
     @Test
     public void testMalformedDocIgnore() throws Exception {
-        String index = "indexmalformed2";
+        String index = "indexmalformed2-" + UUID.randomUUID();
         ElasticSearchConfig config = new ElasticSearchConfig()
                 .setElasticSearchUrl("http://"+container.getHttpHostAddress())
                 .setIndexName(index)
@@ -163,7 +178,7 @@ public class ElasticSearchClientTests {
 
     @Test
     public void testBulkRetry() throws Exception {
-        final String index = "indexbulktest";
+        final String index = "indexbulktest-" + UUID.randomUUID();
         ElasticSearchConfig config = new ElasticSearchConfig()
                 .setElasticSearchUrl("http://"+container.getHttpHostAddress())
                 .setIndexName(index)
@@ -173,35 +188,37 @@ public class ElasticSearchClientTests {
                 .setRetryBackoffInMs(100)
                 .setBulkFlushIntervalInMs(10000);
         ElasticSearchClient client = new ElasticSearchClient(config);
-        client.createIndexIfNeeded(index);
+        assertTrue(client.createIndexIfNeeded(index));
+        try {
+            MockRecord<GenericObject> mockRecord = new MockRecord<>();
+            client.bulkIndex(mockRecord, Pair.of("1", "{\"a\":1}"));
+            client.bulkIndex(mockRecord, Pair.of("2", "{\"a\":2}"));
+            assertEquals(mockRecord.acked, 2);
+            assertEquals(mockRecord.failed, 0);
+            assertEquals(client.totalHits(index), 2);
 
-        MockRecord<GenericObject> mockRecord = new MockRecord<>();
-        client.bulkIndex(mockRecord, Pair.of("1","{\"a\":1}"));
-        client.bulkIndex(mockRecord, Pair.of("2","{\"a\":2}"));
-        Thread.sleep(1000L);
-        assertEquals(mockRecord.acked, 2);
-        assertEquals(mockRecord.failed, 0);
-        assertEquals(client.totalHits(index), 2);
+            ChaosContainer<?> chaosContainer = new ChaosContainer<>(container.getContainerName(), "15s");
+            chaosContainer.start();
 
-        ChaosContainer<?> chaosContainer = new ChaosContainer<>(container.getContainerName(), "15s");
-        chaosContainer.start();
+            client.bulkIndex(mockRecord, Pair.of("3", "{\"a\":3}"));
+            Thread.sleep(5000L);
+            assertEquals(mockRecord.acked, 2);
+            assertEquals(mockRecord.failed, 0);
+            assertEquals(client.totalHits(index), 2);
 
-        client.bulkIndex(mockRecord, Pair.of("3","{\"a\":3}"));
-        Thread.sleep(5000L);
-        assertEquals(mockRecord.acked, 2);
-        assertEquals(mockRecord.failed, 0);
-        assertEquals(client.totalHits(index), 2);
-
-        chaosContainer.stop();
-        client.flush();
-        assertEquals(mockRecord.acked, 3);
-        assertEquals(mockRecord.failed, 0);
-        assertEquals(client.totalHits(index), 3);
+            chaosContainer.stop();
+            client.flush();
+            assertEquals(mockRecord.acked, 3);
+            assertEquals(mockRecord.failed, 0);
+            assertEquals(client.totalHits(index), 3);
+        } finally {
+            client.delete(index);
+        }
     }
 
     @Test
     public void testBulkBlocking() throws Exception {
-        final String index = "indexblocking";
+        final String index = "indexblocking-" + UUID.randomUUID();
         ElasticSearchConfig config = new ElasticSearchConfig()
                 .setElasticSearchUrl("http://"+container.getHttpHostAddress())
                 .setIndexName(index)
@@ -212,39 +229,49 @@ public class ElasticSearchClientTests {
                 .setRetryBackoffInMs(100)
                 .setBulkFlushIntervalInMs(10000);
         ElasticSearchClient client = new ElasticSearchClient(config);
-        client.createIndexIfNeeded(index);
+        assertTrue(client.createIndexIfNeeded(index));
 
-        MockRecord<GenericObject> mockRecord = new MockRecord<>();
-        for(int i = 1; i <= 5; i++) {
-            client.bulkIndex(mockRecord, Pair.of(Integer.toString(i), "{\"a\":"+i+"}"));
-        }
-        // wait bulk flush interval
-        Awaitility.await().untilAsserted( () -> {
-            assertEquals(mockRecord.acked, 5);
+        try {
+            MockRecord<GenericObject> mockRecord = new MockRecord<>();
+            for (int i = 1; i <= 5; i++) {
+                client.bulkIndex(mockRecord, Pair.of(Integer.toString(i), "{\"a\":" + i + "}"));
+            }
+
+            Awaitility.await().untilAsserted(() -> {
+                assertThat("acked record", mockRecord.acked, greaterThanOrEqualTo(4));
+                assertEquals(mockRecord.failed, 0);
+                assertThat("totalHits", client.totalHits(index), greaterThanOrEqualTo(4L));
+            });
+            client.flush();
+            Awaitility.await().untilAsserted(() -> {
+                assertEquals(mockRecord.acked, 5);
+                assertEquals(mockRecord.failed, 0);
+                assertEquals(client.totalHits(index), 5);
+            });
+
+            ChaosContainer<?> chaosContainer = new ChaosContainer<>(container.getContainerName(), "30s");
+            chaosContainer.start();
+            Thread.sleep(1000L);
+
+            // 11th bulkIndex is blocking because we have 2 pending requests, and the 3rd request is blocked.
+            long start = System.currentTimeMillis();
+            for (int i = 6; i <= 15; i++) {
+                client.bulkIndex(mockRecord, Pair.of(Integer.toString(i), "{\"a\":" + i + "}"));
+                log.info("{} index [}", System.currentTimeMillis(), i);
+            }
+            long elapsed = System.currentTimeMillis() - start;
+            log.info("elapsed = {}", elapsed);
+            assertTrue(elapsed > 29000); // bulkIndex was blocking while elasticsearch was down or busy
+
+            Thread.sleep(1000L);
+            assertEquals(mockRecord.acked, 15);
             assertEquals(mockRecord.failed, 0);
-            assertEquals(client.totalHits(index), 5);
-        });
+            assertEquals(client.records.size(), 0);
 
-        ChaosContainer<?> chaosContainer = new ChaosContainer<>(container.getContainerName(), "30s");
-        chaosContainer.start();
-        Thread.sleep(1000L);
-
-        // 11th bulkIndex is blocking because we have 2 pending requests, and the 3rd request is blocked.
-        long start = System.currentTimeMillis();
-        for(int i = 6; i <= 15; i++) {
-            client.bulkIndex(mockRecord, Pair.of(Integer.toString(i), "{\"a\":"+i+"}"));
-            log.info("{} index [}", System.currentTimeMillis(), i);
+            chaosContainer.stop();
+        } finally {
+            client.delete(index);
         }
-        long elapsed = System.currentTimeMillis() - start;
-        log.info("elapsed = {}", elapsed);
-        assertTrue(elapsed > 29000); // bulkIndex was blocking while elasticsearch was down or busy
-
-        Thread.sleep(1000L);
-        assertEquals(mockRecord.acked, 15);
-        assertEquals(mockRecord.failed, 0);
-        assertEquals(client.records.size(), 0);
-
-        chaosContainer.stop();
     }
 
 }

